tfutils/base.py

```diff
@@ -736,78 +801,147 @@ def test_from_params(load_params,
+            assert ld is not None, "No load data found for query, aborting"
+            ld = ld[0]
+            # TODO: have option to reconstitute model_params entirely from
+            # saved object ("revivification")
+            param['model_params']['seed'] = ld['params']['model_params']['seed']
+            cfg_final = ld['params']['model_params']['cfg_final']
+            train_queue_params = ld['params']['train_params']['queue_params']
+
+            (ttarg['validation_targets'],
+             ttarg['queues']) = get_valid_targets_dict(loss_params=None,
+                                                       cfg_final=cfg_final,
+                                                       queue_params=train_queue_params,
+                                                       **param)
+
+            # tf.get_variable_scope().reuse_variables()
+
+            param['load_params']['do_restore'] = True
+            param['model_params']['cfg_final'] = cfg_final
+
+            prefix = param['model_params']['prefix'] + '/'
+            all_vars = variables._all_saveable_objects()
+            var_list = strip_prefix(prefix, all_vars)
+
+            ttarg['dbinterface'] = DBInterface(sess=sess,
+                                               params=param,
+                                               var_list=var_list,
+                                               load_params=param['load_params'],
+                                               save_params=param['save_params'])
+            ttarg['dbinterface'].initialize(no_scratch=True)
+            ttarg['sess'] = sess
+            ttarg['save_intermediate_freq'] = param['save_params'].get('save_intermediate_freq')
+
+        # Convert back to a dictionary of lists
+        params = {key: [param[key] for param in _params]
+                  for key in _params[0].keys()}
+        test_args = {key: [ttarg[key] for ttarg in _ttargs]
+                     for key in _ttargs[0].keys()}
 
         if dont_run:
-            return sess, queues, dbinterface, valid_targets_dict
-
-        save_intermediate_freq = save_params.get('save_intermediate_freq')
-        res = test(sess,
-                   queues,
-                   dbinterface,
-                   valid_targets_dict,
-                   save_intermediate_freq=save_intermediate_freq)
+            return test_args
+
+        res = test(**test_args)
         sess.close()
         return res
 
 
+def train_loop(sess, train_targets, num_minibatches=1, **loop_params):
+    """Define default minibatch training loop.
+
+    A training loop that performs minibatching with `num_minibatches`
+    minibatches.
+
+    Args:
+        sess (tf.Session): Current tensorflow session.
+        train_targets (dict): Target operations to be evaluated by `sess.run`.
+            By default, `base.train_from_params` inserts the following
+            targets to facilitate minibatching:
+            - `grads` (tf.Variable): Stores accumulated gradients.
```

Comment:
should (a) the names of these special ops be more exotic to avoid stepping on them? like "grads" or something? and (b) should this function check that the relevant ops exist?